### ISSUE TITLE
No APIs to explicitly give accessibility focus to React Native elements

### ISSUE THREAD

**Issue by oargaruna** [facebook/react-native#12492]
### Description

Our project is a hybrid of React Native modules and native UI. We are adding accessibility to our App and have hit a roadblock: We cannot explicitly give accessibility focus to an element in the screen so that screen readers, such as Talkback and VoiceOver, can read its accessibility label.

We have some scenarios where we would like to control the accessibility focus after we submit any text in a TextInput.

### Additional Information

* React Native version: 0.41
* Platform: both

**Comment by cannona** [comment]
I haven't tried this, but have you tried setting a ref, and then calling .focus() on that ref? Or perhaps autofocus={true} will work (likely depends on the component type though).

**Comment by oargaruna** [comment]
AFAIK, we can request focus only on `TextInput`, not on an arbitrary `View`.
http://facebook.github.io/react-native/releases/0.43/docs/textinput.html#autofocus

**Comment by johnschimmel** [comment]
looks like new API setAccessibilityFocus to a react component is coming soon in 0.46, <http://facebook.github.io/react-native/releases/0.46/docs/accessibilityinfo.html#setaccessibilityfocus>

**Comment by oargaruna** [comment]
Only for iOS it seems.

**Comment by hramos** [comment]
Hi there! This issue is being closed because it has been inactive for a while. Maybe the issue has been fixed in a recent release, or perhaps it is not affecting a lot of people. Either way, we're automatically closing issues after a period of inactivity. Please do not take it personally!

If you think this issue should definitely remain open, please let us know. The following information is helpful when it comes to determining if the issue should be re-opened:

- Does the issue still reproduce on the latest release candidate? Post a comment with the version you tested.
- If so, is there any information missing from the bug report? Post a comment with all the information required by the [issue template](https://github.com/facebook/react-native/blob/master/.github/ISSUE_TEMPLATE.md).
- Is there a pull request that addresses this issue? Post a comment with the PR number so we can follow up.

If you would like to work on a patch to fix the issue, *contributions are very welcome*! Read through the [contribution guide](http://facebook.github.io/react-native/docs/contributing.html), and feel free to hop into [#react-native](https://discordapp.com/invite/0ZcbPKXt5bZjGY5n) if you need help planning your contribution.

**Comment by shergin** [comment]
@oargaruna Do you know the most comparable iOS API for this?

**Comment by MaxToyberman** [comment]
@oargaruna  I have found a solution for android.
1)
`import {UIManager,findNodeHandle} from 'react-native' `

2) To focus on a View you can use this function:

```
focusOnView = (ref) => {
  UIManager.sendAccessibilityEvent(
     findNodeHandle(ref),
     8  /*  this constant represents  AccessibilityEvent.TYPE_VIEW_FOCUSED in android      */
  )

}
```

3) give a ref to  your view.
 example <View  ref="header"/>

4)then use the function like this:
   focusOnView(this.refs["header"])

don't know why, sometimes you need to add a little timeOut

**Comment by draperunner** [comment]
I am trying to use the `setAccessibilityFocus` API, but I don't understand how it is used. According to the docs, it takes the `reactTag` of an element. What is a reactTag and how do I access an element's reactTag?

I thought I was supposed to pass the `ref` to the element to `setAccessibilityFocus`, so I wrote something like this code:

```
class Header extends React.Component {
    constructor() {
        super()
        this.header = null;
    }

    componentDidMount() {
        if (this.header && this.props.focus) {
            AccessibilityInfo.setAccessibilityFocus(this.header)
        }
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.focus && !this.props.focus) {
            AccessibilityInfo.setAccessibilityFocus(this)
        }
    }

    render() {
        return (
            <View ref={el => { this.header = el }}>
                <Text>
                    { this.props.title }
                </Text>
            </View>
        )
    }
}
```

This explodes and gives me "Converting circular structure to JSON" first, then "Did not get valid calls back from JS: [[40,40,40,40,40,40,40,40,40,40,40,40, ..." with a huge and very unhelpful stacktrace.

**Comment by MaxToyberman** [comment]
@draperunner reactTag is the result of `findNodeHandle(this.header)`

**Comment by oargaruna** [comment]
`AccessibilityInfo.setAccessibilityFocus` works for both iOS and Android. Verified on version 0.51.

**Comment by sospedra** [comment]
@oargaruna this function is not even implemented for Android  ¯\_(ツ)_/¯

**Comment by oargaruna** [comment]
Sorry. I was mistaken. I was switching between the two platforms at the time.

I had to use `AccessibilityInfo.setAccessibilityFocus` for iOS and `UIManager.sendAccessibilityEvent` for Android.
